[Parser] Support "<" unary operator in #if swift() and #if compiler() expressions

danielmartin opened the pull request:
**Swift-evolution proposal:** https://github.com/apple/swift-evolution/pull/877
Resolves #49401.

Until now, only ">=" was supported in #if swift() expressions, for example:

```
#if swift(>=2.1)
#endif
```

This means that if we want to evaluate code only when the language version is
less than a particular version we need to do the following:

```
#if !swift(>=2.1)
#endif
```

An alternative to make this more readable (the "!" can be easily missed in a code
review) is to introduce another supported unary operator, "<". The previous
example could be rewritten like this:

```
#if swift(<2.1)
#endif
```

This commit adds support for that unary operator, along with some tests.

CodaFi commented:
Why not just lift the restriction on operators altogether.

jrose-apple commented:
We don't want to support `>` or `<=` because they don't account for patch versions. That is, if a problem is not fixed in a patch, you'd need to update your condition, and if it *is* fixed in a patch, you can use `>=` or `<` instead.

jrose-apple commented on lib/Parse/ParseIfConfig.cpp:
```diff
@@ -53,6 +53,17 @@ static StringRef extractExprSource(SourceManager &SM, Expr *E) {
+  return UnaryOperator != None && (UnaryOperator.getValue() == ">=" || UnaryOperator.getValue() == "<");
+}
+
+static bool isValidVersion(version::Version &Version, version::Version &ExpectedVersion, StringRef UnaryOperator) {
```
Nitpick: These should be `const` references, and please wrap at 80 columns. (`clang-format` should be set up correctly for this repo.)

danielmartin replied:
Ah, thanks. I've run `git clang-format` to not pollute the diff.

jrose-apple commented on lib/Parse/ParseIfConfig.cpp:
```diff
@@ -53,6 +53,17 @@ static StringRef extractExprSource(SourceManager &SM, Expr *E) {
   return SM.extractText(Range);
 }
 
+static bool isValidPrefixUnaryOperator(llvm::Optional<StringRef> &UnaryOperator) {
```
Nitpick: We import `Optional` into the `swift` namespace like Clang does, so you can just use `Optional<StringRef>` here. Probably no need to pass by reference either.

jrose-apple reviewed (commented):
Actual implementation looks pretty good, just a few nitpicks. Merging needs to wait for the review, of course, or for the Core Team to formally say this doesn't need one (still possible given Chris's positive feedback on the forum thread).

danielmartin committed:
[Parser] Support "<" unary operator in #if swift() expressions

Until now, only ">=" was supported in #if swift() expressions, for example:

```#if swift(>=2.1)
```#endif

This means that if we want to evaluate code only when the language version is
less than a particular version we need to do the following:

```#if !swift(>=2.1)
```#endif

An alernative to make this more readable (the "!" can be easily missed in a code
review) is to introduce another supported unary operator, "<". The previous
example could be rewritten like this:

```#if swift(<2.1)
```#endif

This commit adds support for that unary operator, along with some tests.

rintaro commented on lib/Parse/ParseIfConfig.cpp:
```diff
@@ -448,13 +462,16 @@ class EvaluateIfConfigCondition :
       return thisVersion >= Val;
     } else if ((KindName == "swift") || (KindName == "compiler")) {
       auto PUE = cast<PrefixUnaryExpr>(Arg);
+      auto UDRE = getDeclRefStr(PUE->getFn());
```
Nitpick: `UDRE` (UnresolvedDeclRefExpr) is a little odd here. I suggest `PrefixName` as per `ValidateIfConfigCondition:: visitCallExpr`.

tkremenek commented on test/Parse/ConditionalCompilation/language_version.swift:
```diff
@@ -34,13 +48,21 @@
 #endif
 
-#if swift(">=7.1") // expected-error {{unexpected platform condition argument: expected a unary comparison, such as '>=2.2'}}
+#if !swift(<1000.0)
```
Please also add test cases for `#if compiler`, whose behavior is also changed by this PR.

danielmartin replied:
I noticed that tests for `#if compiler` were missing in the codebase, so I added tests for both operators in a new test file (https://github.com/apple/swift/pull/17960/commits/ccbc3f7dcc8bb41f3ab588b3167f1052875b4095)

tkremenek commented:
SE-0224 has been approved, so this is clear to land once this PR is approved.

danielmartin committed:
Add some tests for #if compiler

benrimmington commented on test/Parse/ConditionalCompilation/compiler.swift:
```diff
@@ -0,0 +1,29 @@
+// RUN: %target-typecheck-verify-swift
```
Could this *also* be run in Swift 4 mode, to check that `#if compiler` is not affected?

```diff
 // RUN: %target-typecheck-verify-swift
+// RUN: %target-typecheck-verify-swift -swift-version 4
```

danielmartin committed:
Run #if compiler tests also in Swift 4 mode

jrose-apple commented:
@tkremenek, I think this is waiting on you.

@swift-ci Please test

tkremenek commented:
Re-running CI against latest master.

@swift-ci test

benrimmington commented:
@tkremenek All checks have passed.
